ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/StreamedQueryResource.java

```diff
@@ -404,6 +510,201 @@ private EndpointResponse handlePullQuery(
+          } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+          }
+          return consumer;
+        }
+
+        @Override
+        public Consumer<byte[], byte[]> getRestoreConsumer(Map<String, Object> map) {
+          return delegateKafkaClientSupplier.getRestoreConsumer(map);
+        }
+
+        @Override
+        public Consumer<byte[], byte[]> getGlobalConsumer(Map<String, Object> map) {
+          return delegateKafkaClientSupplier.getGlobalConsumer(map);
+        }
+
+        @Override
+        public Admin getAdmin(Map<String, Object> config) {
+          return delegateKafkaClientSupplier.getAdmin(config);
+        }
+      };
+    }
+
+    @Override
+    public SchemaRegistryClient getSchemaRegistryClient() {
+      return delegate.getSchemaRegistryClient();
+    }
+
+    @Override
+    public Supplier<SchemaRegistryClient> getSchemaRegistryClientFactory() {
+      return delegate.getSchemaRegistryClientFactory();
+    }
+
+    @Override
+    public ConnectClient getConnectClient() {
+      return delegate.getConnectClient();
+    }
+
+    @Override
+    public SimpleKsqlClient getKsqlClient() {
+      return delegate.getKsqlClient();
+    }
+
+    @Override
+    public KafkaConsumerGroupClient getConsumerGroupClient() {
+      return delegate.getConsumerGroupClient();
+    }
+
+    @Override
+    public void close() {
+      delegate.close();
+    }
+
+    public boolean passed(Map<TopicPartition, Long> endOffsets) {
+      final Set<TopicPartition> finishedPartitions = new HashSet<>();
+      for (SubscriptionState subscriptionState : capturedConsumers) {
+        for (Map.Entry<TopicPartition, Long> entry : endOffsets.entrySet()) {
+          final TopicPartition topicPartition = entry.getKey();
+          try {
+            if (subscriptionState.isAssigned(topicPartition)) {
+              final SubscriptionState.FetchPosition position = subscriptionState.position(topicPartition);
+              if (position.offset == entry.getValue()) {
+                finishedPartitions.add(topicPartition);
+              }
+            }
+          } catch (RuntimeException e) {
+            e.printStackTrace();
+            throw e;
+          }
+        }
+      }
+      return finishedPartitions.equals(endOffsets.keySet());
+    }
+  }
+
+  private EndpointResponse handleStreamPullQuery(
+      final ServiceContext serviceContext,
+      final PreparedStatement<Query> statement,
+      final Map<String, Object> streamsProperties,
+      final CompletableFuture<Void> connectionClosedFuture,
+      final Map<TopicPartition, Long> endOffsets) {
+
+    // stream pull queries always start from earliest.
+    streamsProperties.put("auto.offset.reset", "earliest");
+
+    final ConfiguredStatement<Query> configured = ConfiguredStatement
+        .of(statement, SessionConfig.of(ksqlConfig, streamsProperties));
+
+    if (QueryCapacityUtil.exceedsPushQueryCapacity(ksqlEngine, ksqlRestConfig)) {
```

Comment:
Thanks for pointing this out. I noticed it, but didn't give it much thought. I kind of do think it should be accounted in some way because as implemented, it's a full-blown Streams app for the duration of the query.